tsdb/modifiers.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tsdb
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"sort"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+	"github.com/prometheus/prometheus/tsdb/index"
+	"github.com/prometheus/prometheus/tsdb/tombstones"
+)
+
+type ChangeLogger interface {
+	DeleteSeries(del labels.Labels, intervals tombstones.Intervals)
+	ModifySeries(old, new labels.Labels)
+}
+
+type changeLog struct {
+	w io.Writer
+}
+
+// NewChangeLog creates a change logger writing to the given writer.
+func NewChangeLog(w io.Writer) ChangeLogger {
+	return &changeLog{w: w}
+}
+
+func (l *changeLog) DeleteSeries(del labels.Labels, intervals tombstones.Intervals) {
+	_, _ = fmt.Fprintf(l.w, "Deleted %v %v\n", del.String(), intervals)
+}
+
+func (l *changeLog) ModifySeries(old, new labels.Labels) {
+	_, _ = fmt.Fprintf(l.w, "Relabelled %v %v\n", old.String(), new.String())
+}
+
+// Modifier modifies the index symbols and chunk series before persisting a new block during compaction.
+type Modifier interface {
+	Modify(sym index.StringIter, set storage.ChunkSeriesSet, changeLog ChangeLogger) (index.StringIter, storage.ChunkSeriesSet, error)
+}
+
+// RelabelModifier modifies index via relabeling with changelog support.
+type RelabelModifier struct {
+	relabels []*relabel.Config
+}
```

Comment:
I expected this to go in promtool, since that is the objective of this PR.
Why does it have to go in tsdb?